### DISCUSSION TITLE
Bundle from wic partition

### DISCUSSION THREAD

**Discussion by tcikel** [rauc/meta-rauc#260] [Q&A]
Hello, 

 I have a A/B system with seperate appfs partition. 
My config looks like this:
 
``` ini
[system]
compatible=RaspberryPi4
bootloader=uboot
data-directory=/data/

[keyring]
path=/etc/rauc/ca.cert.pem
 
[slot.rootfs.0]
device=/dev/mmcblk0p2
type=ext4
bootname=A

[slot.rootfs.1]
device=/dev/mmcblk0p3
type=ext4
bootname=B


[slot.appfs.0]
parent=rootfs.0
device=/dev/mmcblk0p5
type=ext4

[slot.appfs.1]
parent=rootfs.1
device=/dev/mmcblk0p6
type=ext4
```

I am trying to build bundle which contains both rootfs and appfs.
I have a simple recipe to fill app partition, it just  copies apps from git  repository into some tmp folder inside yocto build.
I have a wic plugin which gets called later and creates the partition and fills it with data from tmp folder. 

I have bundle recipe where I put the recipe which I am using for appfs partition to its slot
```
RAUC_SLOT_appfs = "apps"
RAUC_SLOT_appfs[fstype] = "ext4"
```
Te build fails on 
`update-bundle.bb depends upon non-existent task do_image_complete in apps.bb`

So from my understanding RAUC expects that the recipe actually builds the artifact as ext4 but the recipe doesnt provide that because its simple and just copies the files. But later the wic plugin creates them so is there any option how to let RAUC know about this?

**Comment by ejoerns** [comment]
The [bundle.bbclass](https://github.com/rauc/meta-rauc/blob/master/classes/bundle.bbclass) expects the image to be generated by an image recipe by default. If it is generated by another sort of recipe, you may want to use another type:

    RAUL_SLOT_appfs[type] = "file"

Then it depends on the `do_deploy` task of the corresponding recipes. The value after `RAUC_SLOT_appfs` is expected to be the BPN of your recipe.

**Reply by tcikel** [reply]
I am not sure if I understand this correctly. The issue is that it is not generated by recipe. I have wks file where I am specifying wic plugin which is used to create partition.
So basically I have to  rework the recipe to create the image right and ditch the wic plugin ?  To be honest I have no idea how to do that, I am still kinda new to Yocto and I couldn't find any good source which actually shows how to create image which is just from partition, all the sources I found are creating just rootfs.

**Reply by ejoerns** [reply]
Do you call wic manually or by using an IMAGE_FSTYPE from within your bitbake run?

**Reply by tcikel** [reply]
I don't call it manually, but also not with IMAGE_FSTYPE. I just include WKS_FILE = "my-wks-file.wks.in" in local.conf, and in wks file I specify wic plugin as source for my appfs partition like this.
`part /app --source app-partition --size 512  --ondisk mmcblk0 --fstype=ext4 --label app_A --align 4096
`
Not sure if this is the right way to do it. I did it like this because in Yocto docs they use it for creating partitioned images which is exactly what I wanted.

EDIT:
Sorry I didnt check my local.confg properly I use wic in IMAGE_FSTYPE. Not sure if it changes anythign though.

**Reply by jagumiel** [reply]
I am dealing with the same problem. Did you finally solve it?

**Reply by JSydll** [reply]
A bit late to the party, but nevertheless: What you can look into is the `build-wic` folder in the workspace of your image recipe - because there will be some image recipe that has (maybe by default?) the `IMAGE_FSTYPE = "wic"` set, which then triggers the build of your partitioned image.

In the `build-wic` folder, you'll find the binary images of each partition. You could deploy these too, so that your bundle recipe can pick them up. 

One way to do that is to extend your image recipe with something like
```
do_copy_wic_partitions() {
    cp -v ${WORKDIR}/build-wic/*.direct.p1 ${IMGDEPLOYDIR}/${IMAGE_BASENAME}${IMAGE_MACHINE_SUFFIX}${IMAGE_NAME_SUFFIX}.appfs.img
}
addtask copy_wic_partitions after do_image_wic before do_image_complete
```

Note however, that you need to use `RAUC_SLOT_appfs[type] = "image"` in this case, as your image recipe does not have a `do_deploy` task, but all other types depend on this. To let the `bundle.bbclass` find your artifact, you can then use `RAUC_SLOT_appfs[fstype] = "appfs.img"` (note the correlation to above nomenclature).
